### PR TITLE
feat: provide better error when ruamel.yaml is not available

### PR DESCRIPTION
**Pull request by appills**
Resolves https://github.com/dupuy/reliabot/issues/15

* note: could not find contribution/style guide, please let me know if anything needs to change

> _edit_: see #23 for CONTRIBUTING.md

### CODE REVIEW

File: reliabot/reliabot.py
Comment (dupuy) [inline reliabot/reliabot.py:52]: `ModuleNotFoundError` is a subclass of `ImportError`, so there's no need to catch both types. The question is whether directing end-users to the installation instructions is likely to be more helpful for the cases of `ImportError` that are _not_ `ModuleNotFoundError` instances.

Based on this [StackOverflow post](https://stackoverflow.com/a/77361812/18829), I think for things like circular imports or errors in the `ruamel.yaml` module itself, the specific exception details would be more useful, as the fix is likely to require code changes, not just proper installation.

Note also in the [pre-commit.ci checks that failed](https://results.pre-commit.ci/run/github/731308823/1704819372.r1pOdxCDRGm3DFCTvEswkA) that you used the Python 2 syntax for multiple exception classes, in Python 3 these need to be parenthesized. It's also helpful to use a more meaningful name for the exception vairable.

```suggestion
except ModuleNotFoundError as module_not_found:
    error_message = ''.join([
        f"{module_not_found}",
```

File: reliabot/reliabot.py
Comment (dupuy) [inline reliabot/reliabot.py:58]: The `sys.exit()` function takes an integer exit code as its argument and doesn't print anything.

The `error(message: str) -> None` function below that will "Report a bug (internal error) in reliabot and exit with status 9." would be a better choice. To use it, the import statements have to be moved after the `error()` function definition.

**Review by dupuy — COMMENTED** [review]
Hey, thanks so much for your contribution. Sorry that I don't have any Contributor docs or guidelines added to this yet, it's a pretty new project.

I've commented on the PR with a suggestion that eliminates a syntax error that prevents a lot of other checks from running at all, so I'll commit that suggestion. If you want to keep working on it to address the other comments, with the [feedback from the pre-commit checks](https://results.pre-commit.ci/run/github/731308823/1704832806.-WnXgAGtTKW9z-UmeRDw9g), that's great. But if not, I'm also happy to do that cleanup myself.

File: reliabot/reliabot.py
Comment (dupuy) [inline reliabot/reliabot.py:58]: To avoid complaints about lowercase for constants (since this is outside any function, as we need to do the import at global level), you could define a proper constant (`RUAMEL_NOT_FOUND`?) before `COMMENT_PREFIX` below, and move the `ruamel.yaml` imports below that (since it needs to be after `error()` anyhow).

File: reliabot/reliabot.py
Comment (dupuy) [reply reliabot/reliabot.py:60]: In PR #20 (for the similar issue #16) I also ended up moving the re2 import down to the top of the "constant" declaration section.

At this point this PR is also a few commits back from the main branch, so I'll try to update the branch so that there's not as much to clean up.

File: reliabot/reliabot.py
Comment (appills) [reply reliabot/reliabot.py]: Changed this to match the formatting you've provided in #20

File: reliabot/reliabot.py
Comment (appills) [reply reliabot/reliabot.py]: Fixed this as well, note that I had to move the import handling even further to account for `DOCTEST_OPT` not being defined - I think a fun task would be to refactor the OPT constants into a separate file, however I am not sure if that is pythonic or a best practice

File: reliabot/reliabot.py
Comment (appills) [reply reliabot/reliabot.py]: Definitely agree with this

File: reliabot/reliabot.py
Comment (appills) [reply reliabot/reliabot.py]: Unresolving in case you want to re-review this

**Comment by appills** [comment]
Went back to remove the `RUAMEL_YAML_NOT_FOUND` constant and remembered why I had the conditional imports - pre-commit.ci keeps hoisting the imports before the ruamel.yaml block. Not sure how you'd like to see this handled.

**Comment by dupuy** [comment]
> pre-commit.ci keeps hoisting the imports

Yes, I've run into unwanted hoisting before and solved it some other way that I don't recall.

It seems the [pre-commit doing this](https://github.com/asottile/reorder-python-imports) has a [`# no-reorder` pragma](https://github.com/asottile/reorder-python-imports?tab=readme-ov-file#using--noreorder) that is probably the best solution in this case. Putting it on its own line just above the three `from ruamel.yaml import`... lines would be a good solution.

File: reliabot/reliabot.py
Comment (dupuy) [inline reliabot/reliabot.py:150]: The missing hyphen seems to be necessary...

```suggestion
# no-reorder
```

Comment (appills) [reply reliabot/reliabot.py]: Just caught this comment, will wait for ci to finish before adding another job to the queue

File: reliabot/reliabot.py
Comment (dupuy) [inline reliabot/reliabot.py:47]: I think anything after the first `# no-reorder` is not hoisted. So the extra comments are probably not needed. But these imports should be moved back down below the `try` block that imports `ruamel.yaml`.

I think that may be the last fix needed. I appreciate your keeping at this. The pre-commit checking can be annoying sometimes, but it also catches a lot of errors, preventing bugs and enforcing consistent style.

Comment (appills) [reply reliabot/reliabot.py:47]: I was employed at one point, I know how much headache ci can cause and prevent 😂 I'll keep bashing at this

> But these imports should be moved back down below the try block that imports ruamel.yaml
Did you mean the try/except block? ci keeps hoisting
    
I did see https://github.com/asottile/reorder-python-imports/blob/3fe08c24a7649551e4f1f77399a233e89e598dec/reorder_python_imports.py#L67 which uses `#noreorder` on `main`

Comment (dupuy) [reply reliabot/reliabot.py:47]: Yes, down between lines 151 and 152. The code you linked to definitely seems to use `noreorder` rather than `no-reorder` as the docs have it. Try moving the `from ruamel.yaml import` lines back down with comments on each line. Hopefully something will stick.

Comment (appills) [reply reliabot/reliabot.py]: I do not believe ` # noreorder` works as desired - I think we want `python-reorder-imports` to skip the lines instead of skipping the reordering. Otherwise, it's bugged.

see: https://github.com/dupuy/reliabot/pull/18/commits/a43a20dd510c996e45463b5a6d3420dfdd636049

I solutioned this to use the conditional import from a previous commit - its not as pretty, but I'll raise an issue with the maintainer to clarify the behavior when I get the time.

**Review by dupuy — APPROVED** [review]
If you want to merge this as-is, with the conditional block, I'm OK with that, so I'm approving this version of the PR. We can see what we can do to fix the problem with the reorder check as a follow-on.

Again, I very much appreciate the extra effort you have put into this. Apropos of your comment about lack of contributor documentation for this project, I have [a PR for that](https://github.com/dupuy/reliabot/pull/23) (still very much in progress), but if you would be willing to review this (as a contributor) when I have gotten further with it, I'd appreciate that too.

Thanks!

File: reliabot/reliabot.py
Comment (dupuy) [reply reliabot/reliabot.py]: 👍

**Comment by appills** [comment]
> If you want to merge this as-is, with the conditional block, I'm OK with that, so I'm approving this version of the PR. We can see what we can do to fix the problem with the reorder check as a follow-on.

I'm totally cool with it, I'd only ask that you squash the commits - I've been working with the github web IDE otherwise I'd do it myself :) do you want to create a task/issue for coming back to the `noreorder` problem, or should we document it via this issue?

> Again, I very much appreciate the extra effort you have put into this. Apropos of your comment about lack of contributor documentation for this project, I have https://github.com/dupuy/reliabot/pull/23 (still very much in progress), but if you would be willing to review this (as a contributor) when I have gotten further with it, I'd appreciate that too.

100% willing to review it once it's ready! Thanks for letting me contribute, I look forward to seeing what else I can do!

File: reliabot/reliabot.py
Comment (dupuy) [inline reliabot/reliabot.py:137]:
```suggestion
```

File: reliabot/reliabot.py
Comment (dupuy) [inline reliabot/reliabot.py:146]:
```suggestion
```

File: reliabot/reliabot.py
Comment (dupuy) [inline reliabot/reliabot.py:150]:
```suggestion
else:
```

**Review by dupuy — COMMENTED** [review]
I poked into the reordering thing a bit more, and figured out two things.

1. The problem wasn't with python-reorder-imports, it was the autopep8 pre-commit check.

2. There was a more elegant way to resolve the problem, using an `else` clause 

It turns out autopep8 also reorders imports (unless they are marked with `# NOQA`), and once it reordered them, python-reorder-imports was leaving them hoisted up. So no need to open a bug on python-reorder-imports, it works as documented and it was just the combination of "too many" pre-commits...

Anyhow, I added suggestions to use the try-except-else approach, it's nicer than a separate conditional.

**Comment by dupuy** [comment]
> I'm totally cool with it, I'd only ask that you squash the commits - I've been working with the github web IDE otherwise I'd do it myself :)

Once all PR conversations are resolved, do you see a green "Squash and merge" button? If not, I should be able to give you merge permissions so that you do.

**Pull request merged**
